### PR TITLE
build user model

### PR DESCRIPTION
**Pull request by baraah-berra**

### CODE REVIEW

File: src/models/user.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+interface UserSchema extends Document {
+  firstname: string;
+  lastname: string;
+  email: string;
+  phone: string;
+  profileImage: string;
+  password: string;
+  role: 'admin' | 'cook' | 'customer';
+  isActive: boolean;
+  address: string;
```
Comment (Khaled6120) [inline src/models/user.ts:21]: address will be of type Address.

File: src/models/user.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  },
+  neighborhood: {
+    type: String,
+    required: [true, 'Neighborhood is required']
+  },
+  streetAddress: {
+    type: String,
+    required: [true, 'Street address is required']
+  }
+});
```
Comment (Khaled6120) [inline src/models/user.ts:41]: no need to create a new schema for the address since the address will be just a nested object in the user model.

**Review by Khaled6120 — CHANGES_REQUESTED** [review]
I just updated the way of implementing the address as we discussed earlier, u can refer to the DB design on Figma for better understanding.

**the updated code will look like this:**


image

image
image

File: src/models/user.ts (hunk 2 above)
Comment (AsliSema) [reply src/models/user.ts:41]: I agree with you. No need to create new schema for address.

File: src/models/user.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
+// Import the necessary modules
+import mongoose, { Document, Schema } from 'mongoose';
+
+// Define the user schema
+interface UserSchema extends Document {
```
Comment (Khaled6120) [inline src/models/user.ts:5]: Rename userSchema `IUser`
and export it
`export interface IUser extends Document {`

File: src/models/user.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  firstname: string;
+  lastname: string;
+  email: string;
+  phone: string;
+  profileImage: string;
+  password: string;
+  role: 'admin' | 'cook' | 'customer';
```
Comment (Khaled6120) [inline src/models/user.ts:12]: `role?: 'admin' | 'cook' | 'customer';`

File: src/models/user.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  isActive: boolean;
```
Comment (Khaled6120) [inline src/models/user.ts:13]: ` isActive?: boolean;`

File: src/models/user.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  address: {
+    city: string;
+    district: string;
+    neighborhood: string;
+    streetAddress: string;
+  };
+}
+
+
+const userSchema: Schema<UserSchema> = new Schema({
```
Comment (Khaled6120) [inline src/models/user.ts:23]: `const userSchema= new Schema<IUser>({`

File: src/models/user.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    neighborhood: {
+      type: String,
+      required: [true, 'Neighborhood is required']
+    },
+    streetAddress: {
+      type: String,
+      required: [true, 'Street address is required']
+    }
+  }
+});
```
Comment (Khaled6120) [inline src/models/user.ts:85]: add `{timestamp:true}`

File: src/models/user.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    required: [true,'Last name is required']
+  },
+  email: {
+    type: String,
+    required: [true, 'Email is required'],
+    unique: true,
+    validate: {
+        validator: (value: string) => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(value),
+        message: 'Invalid email format'
+    }
```
Comment (Khaled6120) [inline src/models/user.ts:39]: no need to implement complicated validator here in the schema, we will add a validation later for that later on.
schema should be simple as much as possible.
so i suggest deleting validate.

File: src/models/user.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+  },
+  phone: {
+    type: String,
+    required: [true, 'Phone number is required'],
+    validate: {
+      validator: (value: string) => /^\d{10}$/.test(value),
+      message: 'Invalid phone number format'
+    }
```
Comment (Khaled6120) [inline src/models/user.ts:47]: no need to implement complicated validator here in the schema, we will add a validation later for that later on.
schema should be simple as much as possible.
so i suggest deleting validate.

File: package.json
Comment (Khaled6120) [inline package.json:1]: Please dont push package.json file in ur next push

**Review by Khaled6120 — CHANGES_REQUESTED** [review]
Thanks Berra so much,
I have just added some comments and I will request changes, please do them they are minor changes.
Also just a reminder, dont push package.json file again, we have already installed all the packages that we need. for the next push, just push the user.ts file.

**Comment by baraah-berra** [comment]
> Thanks Berra so much, I have just added some comments and I will request changes, please do them they are minor changes. Also just a reminder, dont push package.json file again, we have already installed all the packages that we need. for the next push, just push the user.ts file.

Thank you khaled, I reviewed and updated them and pushed user.ts code. I hope it is ready to merge now

**Comment by Khaled6120** [comment]
Berra I requested other changes, maybe u didnt notice them, plz check my review again and make the required changes.
Thanks for ur understanding.

**Review by AsliSema — APPROVED** [review]

**Review by Khaled6120 — APPROVED** [review]

**Pull request closed without merging**